src/v/storage/segment_appender.cc

```diff
@@ -575,16 +572,6 @@ void segment_appender::dispatch_background_head_write() {
                     _opts.priority)
 #pragma clang diagnostic pop
                   .then([this, w](size_t got) {
-                      /*
-                       * the continuation that captured full=true is the end
-                       * of the dependency chain for this chunk. it can be
-                       * returned to cache.
-                       */
-                      if (w->full) {
-                          w->chunk->reset();
-                          internal::chunks().add(w->chunk);
-                      }
```

Comment:
this one was interesting. if `w->full` then we return the chunk to the cache before `w->chunk = nullptr` below. but if not `w->full`, then we still clear it below. so i wonder if `w->full` was always true (or rarely false), because when false it appears that we'd effectively "leak" chunks after clearing `w->chunk` below. but maybe we've just never noticed it.

Reply:
> if not `w->full`, then we still clear it below.

Here we are only "clearing" it from the `struct inflight_write`, but when not full it is still going to be referred to by `segment_appender::_head` and used for later appends.

The  `= nullptr` is not even necessary really, it's just an optimization to return the chunk slightly earlier than it would be otherwise, i.e., after we have written it, but before the `inflight_write` object is destroyed (which may take a while when out of order writes/completions happen as those are destroyed "in order").